# Assignment 2: Income Prediction based on U.S. Census data

Student Name:Yichen Guan

Student ID:*******

# General Info

<b>Due date</b>: Friday, 16 Apr 2021 5pm

<b>Submission method</b>: Canvas submission

<b>Submission materials</b>: completed copy of this iPython notebook

<b>Late submissions</b>: -10% per day (both week and weekend days counted)

<b>Marks</b>: 15% of mark for class 

<b>Materials</b>: See [Using Jupyter Notebook and Python page](https://canvas.lms.unimelb.edu.au/courses/105477/pages/python-and-jupyter-notebooks?module_item_id=2613813) on Canvas (under Modules>Resources) for information on the basic setup required for this class, including an iPython notebook viewer and some handy python packages including Numpy, Scipy, Matplotlib, Scikit-Learn, and Gensim. You can use any Python built-in packages, but do not use any other 3rd party packages (the packages listed above are all fine to use); if your iPython notebook doesn't run on the marker's machine, you will lose marks. <b> You should use Python 3</b>.  

<b>Evaluation</b>: Your iPython notebook should run end-to-end without any errors in a reasonable amount of time, and you must follow all instructions provided below, including specific implementation requirements and instructions for what needs to be printed (please avoid printing output we don't ask for). You should edit the sections below where requested, but leave the rest of the code as is. You should leave the output from running your code in the iPython notebook you submit, to assist with marking. The amount each section is worth is given in parenthesis after the instructions. 

You should be careful to use Python built-in functions and operators when appropriate and pick descriptive variable names that adhere to <a href="https://www.python.org/dev/peps/pep-0008/">Python style requirements</a>. If you think it might be unclear what you are doing, you should comment your code to help the marker make sense of it. While the main focus is on correctness of your methods, you will lose marks if your code is not understandable.

<b>Updates</b>: Any major changes to the assignment will be announced via Canvas. Minor changes and clarifications will be announced on the discussion board; we recommend you check it regularly.

<b>Academic misconduct</b>: For most people, collaboration will form a natural part of the undertaking of this homework, and we encourge you to discuss it in general terms with other students. However, this ultimately is still an individual task, and so reuse of code or other instances of clear influence will be considered cheating. We will be checking submissions for originality and will invoke the University’s <a href="http://academichonesty.unimelb.edu.au/policy.html">Academic Misconduct policy</a> where inappropriate levels of collusion or plagiarism are deemed to have taken place.


## Overview

In this assignment, you will be working with the famous *Adult* a data set containing demographic and income data from the united states in 1994. The dataset provided for this assignment is derived from <a href="http://archive.ics.uci.edu/ml/datasets/Adult">this</a> resource. The data set consists of about 48,000 individuals each characterized through a set of 14 attributes. Your task is to predict whether the individual earns up to \\$ 50,000 a year (<=50K) or more than \\$50,000 per year (>50K).

The attributes are

|ID|Feature Name| Feature Type | Feature Values|
| :-| :-| :-| :-|
|0|age| continuous| |
|1|workclass| categorical | Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, ?|
|2|fnlwgt| continuous| |
|3|education|  categorical |Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool|
|4|education-num| continuous| |
|5|marital-status|  categorical |Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse|
|6|occupation|  categorical |Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces, ?|
|7|relationship|  categorical |Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried|
|8|race|  categorical |White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black|
|9|sex|  categorical |Female, Male|
|10|capital-gain| continuous| |
|11|capital-loss| continuous| |
|12|hours-per-week| continuous| |
|13|native-country|  categorical |United-States, Cambodia, England, Puerto-Rico, Canada, Germany, India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, ?|

You can find out more about the individual attributes / values, and the origin of the data set in the <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names"> data set description</a>  


You will build a number of classifiers to predict the income class based on the attributes above.


#### The following instructions hold for every question in the assignment
- leave the order of instances intact, i.e., do not shuffle the data
- do not change the names or types of variables provided by us in the code cells below
- '?' denotes an unknown value, and you should treat it as just another value for its feature in all tasks below.

## Question 1: Loading and pre-processing the data (1 mark)

You were provide two data files:

**adult.train** contains about 32,000 training instances, one instance per line in comma-separated value (csv) format. Each line contains 14 fields. The first 13 fields correspond to the features listed above, the final field denotes the class label

**adult.test** is formatted exactly like adult.train, and contains about 16,000 further instances for evaluation.


### 1a Read the data [0.5 marks]

First, you will read in the data and create traing features, training labels, test features and test labels. Do not apply any data transformations.

In [82]:
# load data


data = open("adult.train",'r').readlines()
test_data = open("adult.test",'r').readlines()

x_train = []
y_train = []
x_test = []
y_test = []

############################
## your code begins here
############################
for instance in data:
    a = instance.strip().split(",")
    x_train.append(a[:-1])
    y_train.append(a[-1])
    

for test_instance in test_data:
    b = test_instance.strip().split(",")
    x_test.append(b[:-1])
    y_test.append(b[-1])
    
    
############################
## your code ends here
############################

len(x_train), len(y_train), len(x_test), len(y_test), len(x_train[0]), len(x_test[0])
newa =  x_train[0]
print( x_train[0])
print(newa[1])
print(type(x_train))

test_class=" >50K"

print(y_train[0])
print(y_train[9]==test_class)




['39', ' State-gov', ' 77516', ' Bachelors', ' 13', ' Never-married', ' Adm-clerical', ' Not-in-family', ' White', ' Male', ' 2174', ' 0', ' 40', ' United-States']
 State-gov
<class 'list'>
 <=50K
True


<b>For your testing:</b>

In [83]:
assert len(x_train)==len(y_train)==32525
assert len(x_train[0])==len(x_test[0])==14

### 1b: Attribute Types [0.5 marks]

You will create three feature representations, based on the different attribute types (categorical, numeric) in the original *Adult* data.

**Your tasks**

Denote with $I$ the number of training instances; $N$ the number of numeric features in the data set; and $v_f$ the number of possible values for feature $f$.

1. Create a train data set with only numeric features `x_train_num` (size $(I\times N)$); and equivaluently a test data set `x_test_num`
2. Create a train data set with only categorical features *in a 1-hot representation* `x_train_1hot` (size $(I\times\sum_f v_f)$); and equivalently a test data set `x_test_1hot`
3. Create a train data set with both numeric and 1-hot categorical features `x_train_full` (size $I \times N+\sum_f v_f$) where the first $N$ columns represent the numerical features and the remaining columns the categorical features; and equivalently a test data set `x_test_full`

**Note:** You may use classes and functions from ```scikit-learn```.



In [84]:
x_train_num = []
x_test_num = []

x_train_1hot = []
x_test_1hot = []

x_train_full = []
x_test_full = []

################################
### Your code begins here ######
################################

from sklearn.preprocessing import OneHotEncoder   
import numpy as np
####extract features for traning dataset, x_cat_tran1 save untransformate categorical features
x_tran = []
x_cat_tran = []
x_cat_tran1 = []

x_tes =[]
x_cat_test = []
x_cat_test1 = []

for test in x_train:
    for i in range(len(test)):
        if(i==0 or i==2 or i==4 or i==10 or i==11 or i==12 ):
            x_tran.append(test[i])
        if(i==1 or i==3 or i ==5 or i==6 or i==7 or i==8 or i==9 or i==13):
            x_cat_tran.append(test[i])
    
    x_cat_tran1.append(x_cat_tran[0:8])       
    x_train_num.append(x_tran[0:6])
    x_tran.clear()
    x_cat_tran.clear()
    
    
###extract features for test dataset
for instence in x_test:
    for a in range(len(instence)):
        if(a==0 or a==2 or a==4 or a==10 or a==11 or a==12 ):
            x_tes.append(instence[a])
        if(a==1 or a==3 or a==5 or a==6 or a==7 or a==8 or a==9 or a==13):
            x_cat_test.append(instence[a])
    
    x_cat_test1.append(x_cat_test[0:8])    
    x_test_num.append(x_tes[0:6])
    x_tes.clear()
    x_cat_test.clear()

    
############one-hot
enc = OneHotEncoder(handle_unknown='ignore')
enc.fit(x_cat_tran1)
enc.categories_


x_train_1hot=enc.transform(x_cat_tran1).toarray()
x_test_1hot=enc.transform(x_cat_test1).toarray()

x_train_1hot = np.array(x_train_1hot,dtype='i4')
x_test_1hot = np.array(x_test_1hot,dtype='i4')
#print("--------")

#tt=enc.get_feature_names()


#print(tt[0])

####combine

x_train_num = np.array(x_train_num,dtype='i4')
x_test_num = np.array(x_test_num,dtype='i4')

x_train_full = np.hstack((x_train_num ,x_train_1hot))
x_test_full = np.hstack((x_test_num,x_test_1hot))



print("test")
#kkk=x_train_num[0]
#print(type(kkk[0]))
#print(kkk[0])

print(x_train_full[0])

print(x_train_full)
#print(type(x_train_num[0]))

################################
### Your code ends here ########
################################



test
[   39 77516    13  2174     0    40     0     0     0     0     0     0
     1     0     0     0     0     0     0     0     0     0     1     0
     0     0     0     0     0     0     0     0     0     1     0     0
     0     1     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     1     0     0     0     0     0     0     0
     0     1     0     1     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     1     0     0]
[[    39  77516     13 ...      1      0      0]
 [    50  83311     13 ...      1      0      0]
 [    38 215646      9 ...      1      0      0]
 ...
 [    58 151910      9 ...      1      0      0]
 [    22 201490      9 ...      1      0      0]
 [    52 287927      9 ...      1      0      0]]


<b> For your testing</b>

In [85]:
assert len(x_train_1hot[0])==98
assert len(x_train_full[0])==104

## Question 2: A 0-R baseline [0.5 marks]

Implement a zero-r baseline, as introduced in the Evaluation lecture.

In [86]:
from collections import Counter

majority_class = ""
zero_r_predictions = []

############################
## your code begins here
############################
train_array = np.array(y_train)
zero_r = Counter(train_array)
a = zero_r.most_common(1)
hh=a[0]
majority_class = hh[0]

for i in range(len(y_test)):
    zero_r_predictions.append(" <=50K")

        


############################
## your code ends here
############################

print(f"The marority class is: {majority_class}")


The marority class is:  <=50K


## Question 3: Feature selection [2 marks]

In this question you will implement pointwise mutual information (PMI) for feature selection (Question 3a)

In question 3a. you will use the implemented function to create a 1-R classifier based on the single 1-hot attribute  (i.e., categorical feature value) with highest PMI for class ">50K". 
```
argmax_a pmi(a=1,c=">50K")
```
In question 3b. you will apply your 1-R classifier to the test instances, and store your predicted labels in `one_r_predictions`.

<b> You should implement PMI from scratch yourself. You may use native Python libraries like math or numpy to help you, but you may not use existing implementations of PMI.</b>



### 3a. Implement PMI [1 mark]

Implement a function to compute PMI.


In [94]:
'''
PMI: A function to compute the pointwise mutual information between 
all features in the data set and a target class of interest
Input: - features for N input instances
       - labels for N input instances
       - target feature value (default=1)
       - target class (fdefault='>50K')
Output: a dictionary of the form {feature: ppmi} where each 
        feature is denoted by its ID (position in the 1-hot encoded representatoin)

'''

############################
## your code begins here
############################

def pmi(features, labels, tgt_value=1, tgt_class=">50K"):
    
    pmis = {}
    
    features = np.array(features)
    labels = np.array(labels)
    
    features_col = np.size(features,1)
    features_row = np.size(features,0)
####p(label)
    label_counter = Counter(labels)
    label_tgt = label_counter[tgt_class]
    
    p_label = float(label_tgt / features_row)

    for i in range(0,features_col,1):
        k = features[:,i]
        pa_k = k
        counter_same = 0
        for j in range(0,features_row,1):
            t_f = pa_k[j]
            if(t_f == tgt_value and labels[j]==tgt_class):
            #if(pa_k[j] == label[j] and pa_k[j]==tgt_class and label[j]==tgt_class):
                counter_same = counter_same+1
        
        feature_counter = Counter(k)    
        
        feature_tgt = feature_counter[tgt_value]
        
#         print("feature",feature_counter)
        
#         print("feature_tgt",feature_tgt)
        
        p_feature = float(feature_tgt/features_row)
        
#         print("feat prob",p_feature)
        
        ####calculate
#         print("counter:",counter_same)
        p_f_l = float(counter_same/features_row)
    
        p_fml = float(p_feature*p_label)
        if p_f_l == 0:
            pmi_value = float('-inf')
            
        else:
            pmi_value = np.log2(p_f_l/p_fml)
            pmis[i] = pmi_value 
            
            
        
#         print("p_f_l",p_f_l)
#         print("p_fml",p_fml)
  
#         pmi_value = np.log2(p_f_l/p_fml)
# #         print("pmi :",i,"=",pmi_value)
#         pmis[i] = pmi_value   
             
    return pmis


############################
## your code ends here
############################


<b>For your testing:</b>

In [95]:
test_features = [[1,1], [1,0], [0,1], [0,0]]
test_labels = [1,1,0,0]
test_pmi = pmi(test_features, test_labels, tgt_class=1)

# test_n_features = x_train_1hot
# test_n_labels = y_train
# test_pmi1 = pmi(test_n_features, test_n_labels, tgt_class=" >50K")

# where the index 0 refers to "feature 1" and index 1 refers to "feature 2"
assert test_pmi[0]==1.0
assert test_pmi[1]==0.0



### 3b Create 1-R baseline [1 mark]

- Apply your PMI function to the *1_hot feature representation* of training data, and determine the (i) 1-hot feature with *highest* PMI with class '>50K' () and (ii) 1-hot feature with *lowest* (most negative) PMI with class '>50K'. Store the name (string) of the corresponding highest/lowest PMI features in `highest_pmi_feature_name` and `lowest_pmi_feature_name`, respectively.

- The feature with *highest* PMI will consitute your 1-R predictor, which you should use to predict the class labels for the test set (`one_r_predictions`)

In [97]:
one_r_predictions = []

highest_pmi_feature_name = "" # feature with highest PMI
lowest_pmi_feature_name = "" # feature with lowest PMI

############################
## your code begins here
############################

pmi_dic = pmi(x_train_1hot, y_train, tgt_class=" >50K")
max_key = max(pmi_dic, key=pmi_dic.get)
min_key = min(pmi_dic, key=pmi_dic.get)
#print(max_key)
#print(min_key)

tt=enc.get_feature_names()
highest_pmi_feature_name = tt[max_key]
lowest_pmi_feature_name = tt[min_key]


for one_r in x_test_1hot:
    if one_r[max_key] == 1:
        one_r_predictions.append(" >50K")
    if one_r[max_key] == 0:
        one_r_predictions.append(" <=50K")
        
#print(len(one_r_predictions))
        
# print(tt[max_key])
# print(tt[min_key])
############################
## your code ends here
############################
print(one_r_predictions[:10])
print(f"The feature with highest PMI for the class '>50K' is: {highest_pmi_feature_name}")
print(f"The feature with lowest PMI for the class '>50K' is: {lowest_pmi_feature_name}")

[' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K', ' <=50K']
The feature with highest PMI for the class '>50K' is: x1_ Doctorate
The feature with lowest PMI for the class '>50K' is: x3_ Priv-house-serv


## Question 4: Naive Bayes [3 marks]

We will construct three Naive Bayes classifiers

1. One for instances with only 1-hot encoded (binary) categorical attributes.
2. One for instances with only numerical attributes.
3. One for instances with the full set of numerical *and* categorical attributes, ensuring that your classifier computes posterior class probabilities $p(y|x)$ as $p(x|y)p(y)$.

For each classifier, you will 
1. Train it on the training set
2. Use the models to predict labels for the test set


### 4a Implementing the Naive Bayes classifiers [1.5 marks]

Implement three functions which train a NB classifier given the specified input feature types and predict labels for a given test set.

You may add additional quantities to your `return` statements.

<b> You may (and are, indeed, encouraged) to, use the existing NB implementations from `sklearn`. You should use the default parameterizations of these algorithms.</b> 
    
If you choose to implement your classifiers from scratch, please use Laplace smoothing (alpha=1) for the categorical feature NB.


In [98]:

##################################
## your code begins here
##################################

from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import CategoricalNB
from sklearn.preprocessing import KBinsDiscretizer

def nb_binary_features(train_features, train_labels, test_features):
    predictions = []
    
    BernNB = BernoulliNB(alpha=1.0)
    BernNB.fit(train_features,train_labels)
    predictions = BernNB.predict(test_features)
    
    return predictions


def nb_continuous_features(train_features, train_labels, test_features):
    predictions = []
    
    GausNB = GaussianNB()
    GausNB.fit(train_features,train_labels)
    
    predictions = GausNB.predict(test_features)
    
    return predictions


def nb_full(train_features, train_labels, test_features):  
#     print("---star")
#     print(train_features)
    predictions = []
    
    ###deep copy
    data_or_train_fe = train_features.copy()
    data_or_test_fe = test_features.copy()
    
    ###data discretization
    kdb = KBinsDiscretizer(strategy="uniform",n_bins=5,encode="ordinal")
    data_disc_train = kdb.fit_transform(data_or_train_fe[:,0:6])
    data_disc_test = kdb.fit_transform(data_or_test_fe[:,0:6])
    
    data_or_train_fe[:,0:6] = data_disc_train
    data_or_test_fe[:,0:6] = data_disc_test
#     print("-------")
#     print(kdb.bin_edges_)
#     print(data_disc_train)
 
    #####model
    CateNB = CategoricalNB(alpha=1.0)
    CateNB.fit(data_or_train_fe,train_labels)
    predictions = CateNB.predict(data_or_test_fe)
    return predictions




############################
## your code ends here
############################


### 4b Apply your classifiers to the data sets you created in Questions 1 and 3b [0.5 marks]

...namely, the

1. 1-hot categorical features `x_{train,test}_1hot`
2. numerical features `x_{train,test}_num`
3. combined numerical and 1-hot categorical features `x_{train,test}_full`


In [99]:
from collections import Counter

############################
## your code begins here
############################


categorical_nb_predictions = nb_binary_features(x_train_1hot,y_train,x_test_1hot)

numeric_nb_predictions = nb_continuous_features(x_train_num,y_train,x_test_num)
# print("after1")
# print(x_train_full)
full_nb_predictions = nb_full(x_train_full,y_train,x_test_full)
# print("after")
# print(x_train_full)

############################
## your code ends here
############################
print(f"Numerical NB predicted class distribution \t {Counter(numeric_nb_predictions)}")
print(f"Categorical NB predicted class distribution\t {Counter(categorical_nb_predictions)}")
print(f"Full NB predicted class distribution\t {Counter(full_nb_predictions)}")


Numerical NB predicted class distribution 	 Counter({' <=50K': 14440, ' >50K': 1822})
Categorical NB predicted class distribution	 Counter({' <=50K': 10046, ' >50K': 6216})
Full NB predicted class distribution	 Counter({' <=50K': 10593, ' >50K': 5669})



### 4c Explain your Naive Bayes classifier implementation on the full attribute set (numerical and binary features). [1 mark]

Please limit your answer to 2-3 sentences


Some of the numerical features have a strong correlation with binary features, such as education-num and education, high education level will take a long education time. So I discretization of numerical features and use Categorical Naive Bayes to implement classifiers. The reason why use Categorical Naive Bayes is after discretization, all features are categorical type.

## Question 5: Logistic Regression [2.5 marks]

Apply a Logistic Regression classifier to the full training data set (`x_{train,test}_full`)

<b> Use the existing implementation in sklearn with default parameters. </b>

### 5a The Logistic Regression classifier [0.5 marks]

In [100]:
from sklearn.linear_model import LogisticRegression
import heapq
lr_predictions = []

############################
## your code begins here
############################

max_value_list = []

def LR_Class_features(train_features, train_labels, test_features):
    predictions = []
    
    lr = LogisticRegression()
    lr.fit(train_features,train_labels)
    
    predictions = lr.predict(test_features)
    
######get the first two max coefficient for 5b.a
    check_ft_max = lr.coef_[0]
    check_ft_max1 = check_ft_max.copy()
    max_num_index_list=heapq.nlargest(2, range(len(check_ft_max1)), check_ft_max1.take)
    first_index = max_num_index_list[0]
    second_index = max_num_index_list[1]
    f_v = check_ft_max[first_index]
    s_v = check_ft_max[second_index]
    
    ####in the PMI ,feature 17 is highest value, feature 20 is lowest value
    print("coefficient value of highest pmi",check_ft_max[17])
    print("coefficient value of lowest pmi",check_ft_max[20])
    
    
    
    return predictions,max_num_index_list,f_v,s_v


lr_predictions,max_value_list,f_coef,s_coef = LR_Class_features(x_train_full,y_train,x_test_full)

max_key_coef = max_value_list[0]
max_key1_coef = max_value_list[1]


num_feature_name = ['age','fnlwgt','education-num','capital-gain','capital-loss','hours-per-week']
num_feature_name.extend(tt)

#print(num_feature_name)
#print(tt)
highest_coef_feature_name = num_feature_name[max_key_coef]
secHigh_coef_feature_name = num_feature_name[max_key1_coef]

print("--------")
print(max_value_list)
print("the high coefficient value is ",f_coef )
print("the second high coefficient value is ",s_coef )

print("the high coefficient feature is ",highest_coef_feature_name )
print("the second high coefficient feature is ",secHigh_coef_feature_name )

############################
## your code ends here
############################

print(f"LR predicted class distribution\t {Counter(lr_predictions)}")

coefficient value of highest pmi -1.1964256579555634e-05
coefficient value of lowest pmi -8.401180992124138e-06
--------
[4, 31]
the high coefficient value is  0.000780962890351613
the second high coefficient value is  0.0003772312253067222
the high coefficient feature is  capital-loss
the second high coefficient feature is  x2_ Married-civ-spouse
LR predicted class distribution	 Counter({' <=50K': 14837, ' >50K': 1425})


### 5b Inspect the learnt coefficients [2 marks]

**Part 1**
Inspect the coefficients (weights) learnt by the classifier, and compare them to the PMI values in Question 1. You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 5b.

**Part 2**
What does a high coefficient for a feature imply? For which two features does your model learn the largest coefficients? Compare your answer to the highest PMI features you detected in Q3.

The high coefficient means that when a feature appears, the odds of the event occurring will increase. 
Features of capital-loss and Married-civ-spouse are have the largest coefficients from the logistic regression model.

## Question 6: Evaluation [1 mark]

We will evaluate our baselines and classifiers on the instances in the test set. 

Compute 
- accuracy
- macro-averaged F1 score 

for the two baselines, the three NB classifiers and the LR classifier.

**You may use existing implementations and/or Python libraries like numpy, scipy or sklearn.**

In [101]:
zero_r_acc = 0
zero_r_f1 = 0

one_r_acc = 0
one_r_f1 = 0

num_nb_acc = 0
num_nb_f1 = 0

cat_nb_acc = 0
cat_nb_f1 = 0

full_nb_acc = 0
full_nb_f1 = 0

lr_acc = 0
lr_f1 = 0

############################
## your code begins here
############################
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

######zero r
zero_r_acc = accuracy_score(y_test, zero_r_predictions)
zero_r_f1 = f1_score(y_test, zero_r_predictions, average='macro')

#### one r
one_r_acc = accuracy_score(y_test, one_r_predictions)
one_r_f1 = f1_score(y_test, one_r_predictions, average='macro')


#### num BY
num_nb_acc =accuracy_score(y_test, numeric_nb_predictions)
num_nb_f1 = f1_score(y_test, numeric_nb_predictions, average='macro')


#### categrate BY
cat_nb_acc =accuracy_score(y_test, categorical_nb_predictions)
cat_nb_f1 = f1_score(y_test, categorical_nb_predictions, average='macro')


#### FULL BY
full_nb_acc =accuracy_score(y_test, full_nb_predictions)
full_nb_f1 = f1_score(y_test, full_nb_predictions, average='macro')

##### LR
lr_acc = accuracy_score(y_test, lr_predictions)
lr_f1 = f1_score(y_test, lr_predictions, average='macro')


############################
## your code ends here
############################

print(f"Zero R\t\tAccuracy: {round(zero_r_acc, 2)}\tMacro F1: {round(zero_r_f1, 2)}")
print(f"One  R\t\tAccuracy: {round(one_r_acc, 2)}\tMacro F1: {round(one_r_f1, 2)}")
print(f"NB Num \t\tAccuracy: {round(num_nb_acc, 2)}\tMacro F1: {round(num_nb_f1, 2)}")
print(f"NB Cat \t\tAccuracy: {round(cat_nb_acc, 2)}\tMacro F1: {round(cat_nb_f1, 2)}")
print(f"NB Full \tAccuracy: {round(full_nb_acc, 2)}\tMacro F1: {round(full_nb_f1, 2)}")
print(f"LR \t\tAccuracy: {round(lr_acc, 2)}\tMacro F1: {round(lr_f1, 2)}")

Zero R		Accuracy: 0.76	Macro F1: 0.43
One  R		Accuracy: 0.77	Macro F1: 0.46
NB Num 		Accuracy: 0.8	Macro F1: 0.64
NB Cat 		Accuracy: 0.76	Macro F1: 0.72
NB Full 	Accuracy: 0.79	Macro F1: 0.75
LR 		Accuracy: 0.8	Macro F1: 0.63


## Question 7: Discussion [5 marks]

Critically analyze the performance of the models by answering the following questions. 

**(a)** The three Naive Bayes (NB) classifiers lead to different performance. Which of the three NB classifiers performs best, and why do you think it is the case? **[1 mark]**

**(b)** There is a systematic difference between Accuracy and F1 score, across all models. Describe and explain the difference in the context of the data set. **[1 mark]**

**(c)** Assume that your best classifier will be deployed in a bank to decide whether a customer will be granted a loan, or not. A loan will be granted if the classifier predicts an income of ">50K", othewise it won't. The bank manager wants to avoid falsely granting a loan to an applicant with insufficient income by all means. Can you reassure the bank manager with the evaluation results in Q7 that your classifier is adequate? If not, describe an alternative evaluation metric and explain why it is more appropriate.

**(d)** Do you observe a clear difference in performance between NB and Logistic Regression (LR) in Q7? Referring to the number of parameters and assumptions underlying each model, provide one reason why NB might outperform LR, and one reason why LR might outperform NB. In your answer, refer back to the *Adult* data set used in this assignment. **[1 mark]**

**(e)** Assume that you have access to a (hypothetical) infinitely large U.S. census database (i.e., an enormous version of the *Adult* data set used in this assignment). Is it true that Logistic Regression, but not Naive Bayes, will achieve perfect test set performance? Why (not)? **[1 mark]**

<b>We expect a maximum of 2-3 sentences per questinon.

In [102]:
######test for answer question 7b
print(Counter(y_train))
print(Counter(y_test))


# q=np.array(x_train)

# print(Counter (q[:,1]))
# print(Counter (q[:,6]))

Counter({' <=50K': 24684, ' >50K': 7841})
Counter({' <=50K': 12419, ' >50K': 3843})




(a) NB FULL is the best classifier. Because the Macro F1 value is the biggest compare with the others which mean this model has a strong ability to distinguish both ">50k" and  " <50k". Reduce the impact of uneven data distribution.


(b)The uneven class distribution make the Accuracy is not good for evaluate the model. In the training data set, "<50K" has 24684 instances and  ">50K" only has 7841. that is also a reason of why some model has higher accuracy but lower F1 score. 


(c)NO, because Macro F1 average the Precision and Recall. For this situation, the bank manager wants to avoid falsely granting a loan to people who have less than 50k income. So it is important to predict whether "<50" is accurate, the Precision of the model should have a higher weight to evaluate the classifier model.


(d)
NB outperform LR: the Naive Bayes can handle both continuous and discrete data but logistic regression can not handle continuous data.The data needs to be discretized and normalized in advance to reduce the gradient. If the continuous data is not discretized, the accuracy of the logistic regression model will be disturbed by outliers.

LR outperform NB: base on the assumption of Naive Bayes, the feature should be independent of each other and all features should have the same weight. However observe the feature of the data set, some of them have relevance. For example, "eduacation" and "education-num", "capital-gain" and "hours-per-week". Non-independent features have little effect on logistic regression.



(e)Yes, it is. because the assumption of Naive Bayes is independence among features,  when the dataset is infinitely large, the variance will be increased, make the naive Bayes perform less accurately.

## Final thoughts ...

Do you think it is ethical and safe to build and deploy income-level classifiers based on the set of attributes provided with the data set? Could such a classifier do harm if not developed, tested and deployed carefully?

When predicting the income of a person, should demographic attributes like 'gender' or 'race' even be considered? Due to historical events there are systematic income differences between men/women and people of different ethnicities. But no doubt these are *correlations* and not factors to be used to *predict* income in the future. 

But ML models are only as good as their input data: a naively developed ML algorithm will reflect, and often amplify, biases in the data.

In the last 1-2 weeks of this subject we will discuss ethics in machine learning and look at questions such as
- historical artifacts which lead to biased data sets, and hence potentially biased ML algorithms
- how to determine whether an algorithm is indeed biased
- how to mitigate bias in the data and in algorithms

